src/common/src/util/stream_cancel.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright 2023 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A library inspired by [stream-cancel](https://github.com/jonhoo/stream-cancel). This
+//! difference is that it allows you to send a value to original stream when cancelled, this is
+//! useful in several cases, for example when cancel query, it can be used to concat an cancel
+//! error to original stream.
+
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+use std::task::Poll;
+
+use futures::channel::oneshot::{channel, Receiver, Sender};
+use futures::stream::poll_fn;
+use futures::{Stream, StreamExt};
+
+pub struct Trigger {
+    sender: Sender<()>,
+}
+
+impl Trigger {
+    /// Used to trigger cancelling of stream.
+    pub fn abort(self) {
+        if self.sender.send(()).is_err() {
+            tracing::debug!("Trigger receiver closed, ignoring...");
+        }
+    }
+}
+
+pub struct Tripwire<R> {
```

Comment:
Could you please add some rustdocs for these public stuff?

Reply:
I think it's better to add an example to doc.